scheduledJobs/giveApples.js

```diff
@@ -0,0 +1,332 @@
+                // i.e. homework logbook
+                setLogbookUserIds(message, userIds);
+                filteredMembers.forEach((m) => membersNeedingApples.push(m));
+            }
+        });
+    };
+
+    const removeRollCallRole = async (member) => {
+        // roll call role removed from students given a '🍏'
+        const hasRollCollRole = member.roles.cache.has(rollCallRoleId);
+        const hasGreenAppleRole = member.roles.cache.has(greenAppleRoleId) || usersWithGreenAppleAdded.has(member.id);
+        if (hasRollCollRole && hasGreenAppleRole) {
+            await member.roles.remove([rollCallRoleId]);
+            usersWithRollCallRemoved.add(member.id);
+        }
+    };
+
+    const addActiveStudentRole = async (member) => {
+        // active student role added to students given a '🍎'
+        const hasActiveStudentRole = member.roles.cache.has(activeStudentRoleId);
+        const hasRedAppleRole = member.roles.cache.has(redAppleRoleId) || usersWithRedAppleAdded.has(member.id);
+        if (!hasActiveStudentRole && hasRedAppleRole) {
+            await member.roles.add([activeStudentRoleId]);
+            usersWithActiveStudentAdded.add(member.id);
+        }
+    };
+
+    const addAppleRoles = async () => {
+        const bothRoles = [greenAppleRoleId, redAppleRoleId];
+        if (logbookUserIds.length === 0 || membersNeedingApples.length === 0) {
+            return;
+        }
+
+        const userFrequency = {};
+        for (let i = 0; i < logbookUserIds.length; i++) {
+            userFrequency[logbookUserIds[i]] = userFrequency[logbookUserIds[i]] ? userFrequency[logbookUserIds[i]] + 1 : 1;
+        }
+
+        await Promise.all(membersNeedingApples.map(async (member) => {
+            if (userFrequency[member.id] >= 2) {
+                await member.roles.add(bothRoles);
+                usersWithGreenAppleAdded.add(member.id);
+                usersWithRedAppleAdded.add(member.id);
+            } else if (!member.roles.cache.has(greenAppleRoleId)) {
+                await member.roles.add([greenAppleRoleId]);
+                usersWithGreenAppleAdded.add(member.id);
+            } else {
+                await member.roles.add([redAppleRoleId]);
+                usersWithRedAppleAdded.add(member.id);
+            }
+            await removeRollCallRole(member);
+            await addActiveStudentRole(member);
+        }));
+    };
+
+    const addOnlyRedAppleRole = async () => {
+        if (membersNeedingOnlyRedApples.size === 0) {
+            return;
+        }
+        const members = Array.from(membersNeedingOnlyRedApples.values());
+        await Promise.all(members.map(async (member) => {
+            // must run after addAppleRoles to ensure cache has been updated
+            if (!member.roles.cache.has(redAppleRoleId)) {
+                await member.roles.add([redAppleRoleId]);
+                usersWithRedAppleAdded.add(member.id);
+            }
+            await removeRollCallRole(member);
```

Comment:
Do we need to call this if we assign only 🍎 that doesnt affect the roll call? It probably doesnt affect anything except the performance? Just wondering! (omg i hate giving comments like this i always feel like offending that person, im sorry Alex i love you to bits 😭💜)

Reply:
no no not offensive at all lovely!! <3 

in `removeRollCallRole` we check if they have a green apple or not before trying to remove it. i sent the red apple folks here just in case, because i wanted to catch anyone who may not have had that role removed previously but are still eligible to have the role removed. just one less role removal for staff to worry about. 

but we can totally remove it though if it is confusing to have here. i'm not _super_ worried about performance since it is a nightly job, but it should resolve super quickly if they aren't passing the check to have the role removed. 🤞